### PR TITLE
ffmpeg: reallocate output buffer dynamically

### PR DESCRIPTION
**Pull request by equeim**
With FFmpeg we can't determine size of output buffer ahead of time for all codecs, so we need to reallocate it when needed instead of simply failing.

### CODE REVIEW

**Comment by tonihei** [comment]
Thanks for the PR! I added a few comments for improvement ideas.

**Comment by equeim** [comment]
> Thanks for the PR! I added a few comments for improvement ideas.

Where? 🤔

File: libraries/decoder_ffmpeg/src/main/jni/ffmpeg_jni.cc [hunk 1]
```diff
@@ -107,6 +117,17 @@ void releaseContext(AVCodecContext *context);
 jint JNI_OnLoad(JavaVM *vm, void *reserved) {
   JNIEnv *env;
   if (vm->GetEnv(reinterpret_cast<void **>(&env), JNI_VERSION_1_6) != JNI_OK) {
+    LOGE("JNI_OnLoad: GetEnv failed");
+    return -1;
+  }
+  jclass clazz = env->FindClass("androidx/media3/decoder/ffmpeg/FfmpegAudioDecoder");
```
[the remark this answers is not in the document]
Comment (equeim) [reply libraries/decoder_ffmpeg/src/main/jni/ffmpeg_jni.cc:125]: Done.

File: libraries/decoder/src/main/java/androidx/media3/decoder/SimpleDecoderOutputBuffer.java [hunk 2]
```diff
@@ -49,6 +50,18 @@ public ByteBuffer init(long timeUs, int size) {
     return data;
   }
 
+  public ByteBuffer grow(int newSize) {
```
[the remark this answers is not in the document]
Comment (equeim) [reply libraries/decoder/src/main/java/androidx/media3/decoder/SimpleDecoderOutputBuffer.java:60]: Done.

File: libraries/decoder_ffmpeg/src/main/jni/ffmpeg_jni.cc [hunk 3]
```diff
@@ -320,15 +351,22 @@ int decodePacket(AVCodecContext *context, AVPacket *packet,
...
       context->opaque = resampleContext;
     }
-    int inSampleSize = av_get_bytes_per_sample(sampleFormat);
+
     int outSampleSize = av_get_bytes_per_sample(context->request_sample_fmt);
     int outSamples = swr_get_out_samples(resampleContext, sampleCount);
     int bufferOutSize = outSampleSize * channelCount * outSamples;
     if (outSize + bufferOutSize > outputSize) {
-      LOGE("Output buffer size (%d) too small for output data (%d).",
+      LOGE("Output buffer size (%d) too small for output data (%d), reallocating buffer.",
```
[the remark this answers is not in the document]
Comment (equeim) [reply libraries/decoder_ffmpeg/src/main/jni/ffmpeg_jni.cc]: Done.

File: libraries/decoder_ffmpeg/src/main/jni/ffmpeg_jni.cc [hunk 4]
```diff
@@ -320,15 +351,22 @@ int decodePacket(AVCodecContext *context, AVPacket *packet,
...
            outputSize, outSize + bufferOutSize);
-      av_frame_free(&frame);
-      return AUDIO_DECODER_ERROR_INVALID_DATA;
+      outputSize = outSize + bufferOutSize;
+      outputBuffer = growBuffer(outputSize);
+      if (outputBuffer) {
+        LOGE("Reallocated output buffer.");
```
[the remark this answers is not in the document]
Comment (equeim) [reply libraries/decoder_ffmpeg/src/main/jni/ffmpeg_jni.cc]: Done.

File: libraries/decoder_ffmpeg/src/main/java/androidx/media3/decoder/ffmpeg/FfmpegAudioDecoder.java [hunk 5]
```diff
@@ -107,8 +109,10 @@ protected FfmpegDecoderException decode(
     }
     ByteBuffer inputData = Util.castNonNull(inputBuffer.data);
     int inputSize = inputData.limit();
-    ByteBuffer outputData = outputBuffer.init(inputBuffer.timeUs, outputBufferSize);
-    int result = ffmpegDecode(nativeContext, inputData, inputSize, outputData, outputBufferSize);
+    outputBuffer.init(inputBuffer.timeUs, outputBufferSize);
+    Assertions.checkNotNull(outputBuffer.data);
```
[the remark this answers is not in the document]
Comment (equeim) [reply libraries/decoder_ffmpeg/src/main/java/androidx/media3/decoder/ffmpeg/FfmpegAudioDecoder.java]: Ok, but then IDE paints `outputBuffer.data` on the next line yellow.

File: libraries/decoder_ffmpeg/src/main/java/androidx/media3/decoder/ffmpeg/FfmpegAudioDecoder.java [hunk 6]
```diff
@@ -135,11 +139,19 @@ protected FfmpegDecoderException decode(
...
     }
-    outputData.position(0);
-    outputData.limit(result);
+    outputBuffer.data.position(0);
+    outputBuffer.data.limit(result);
     return null;
   }
 
+  // Called from native code
+  @Keep
```
[the remark this answers is not in the document]
Comment (equeim) [reply libraries/decoder_ffmpeg/src/main/java/androidx/media3/decoder/ffmpeg/FfmpegAudioDecoder.java]: Done.

**Comment by tonihei** [comment]
Thanks! I'll now import this PR internally and you may see some further clean-up commits appear here.

**Comment by Tolriq** [comment]
@tonihei Your commit https://github.com/androidx/media/commit/8750ed8de6469dc818007f2eb254df9ddbd52cc5 breaks this.

By using a variable to hold the buffer you keep a reference to the old buffer at the end of the function.

So that call `outputData.limit(result);` fails as the result value can be related to the new buffer size but outputData is still the lowersize buffer and you have `Caused by: java.lang.IllegalArgumentException: newLimit > capacity: (73728 > 65535)` errors.

**Comment by tonihei** [comment]
Oh yes, sorry, that was clearly too much "clean-up" :)  Will send a fix commit straight-away.

**Comment by tonihei** [comment]
Should be fixed by https://github.com/androidx/media/commit/ae6f83d298424bd405372803bb8b206fc95a2d0f

**Comment by Tolriq** [comment]
@tonihei yes the fix works but the keep rules are not working :( 
We need to also keep `SimpleDecoderOutputBuffer` else it's renamed and eventually merged and the `GetMethodID` fails.

**Comment by Tolriq** [comment]
@tonihei In this case just compile any app with ffmpeg with R8 and the extension will fail to load in the `JNI_OnLoad` so no need for a file to repro.

And I can send you a file that repro this by mail if you want for future reference too.

For the PR even after all those years I'm still far from a proguard/R8 expert.

So `-keep class androidx.media3.decoder.SimpleDecoderOutputBuffer { *; } ` Do work and fix this, but there's probably a more targeted solution to this.

As bonus questions since there's movement in that extension: Is there any chance that https://github.com/androidx/media/pull/707 be considered? And if yes is there any interest in a PR that adds resampling to ffmpeg extension so that hi res media that Android can't play can properly be downsampled to the max supported sample rate ?

**Comment by tonihei** [comment]
Yes, if you could send a file that needs this logic to android-media-github@google.com with "Issue #746" in the subject, that would be great so that I can just test it myself.

**Comment by Tolriq** [comment]
Done please confirm when I can delete the file.

**Comment by Tolriq** [comment]
And just in case I was not clear enough @tonihei ffmpeg extension in main is broken for any file in current state if you build with R8 as it fails to load the extension at all.

**Comment by tonihei** [comment]
@Tolriq I can't reproduce the R8 failure I'm afraid. I'm testing on the latest `main` branch demo app in release mode with your example file. And everything works as expected (including the automated buffer resizing).

The existing proguard rule seems to ensure that the `SimpleDecoderOutputBuffer` class is kept already. I added
```
-whyareyoukeeping class androidx.media3.decoder.SimpleDecoderOutputBuffer
```
to get an explanation for why this class is kept in my build and the output directly references the existing rule:
```
androidx.media3.decoder.SimpleDecoderOutputBuffer
|- is referenced from:
|  java.nio.ByteBuffer androidx.media3.decoder.ffmpeg.FfmpegAudioDecoder.growOutputBuffer(androidx.media3.decoder.SimpleDecoderOutputBuffer,int)
|- is referenced in keep rule:
|  [...]/media/libraries/decoder_ffmpeg/buildout/intermediates/consumer_proguard_dir/release/lib0/proguard.txt:9:1
```

Could you check that the Ffmpeg proguard rules are included in your build? (maybe you need to clean caches/builds and try again?)

**Comment by Tolriq** [comment]
@tonihei The rules are included I'm using R8 with full mode and version "8.2.34" 

In that mode the SimpleDecoderOutputBuffer is renamed and merged leading to the issue.

Maybe it's an R8 issue and the fact that the class is passed as a param should ensure it's kept, but from my understanding of R8/Proguard there's no guarantee about that with the current rule.

The rule

```
-keep class androidx.media3.decoder.ffmpeg.FfmpegAudioDecoder {
  private java.nio.ByteBuffer growOutputBuffer(androidx.media3.decoder.SimpleDecoderOutputBuffer, int);
}
```

Just implies that the class and function is kept nothing about the parameters.

And in the code we do reflection with
```
  growOutputBufferMethod =
      env->GetMethodID(clazz, "growOutputBuffer",
                       "(Landroidx/media3/decoder/"
                       "SimpleDecoderOutputBuffer;I)Ljava/nio/ByteBuffer;");
```

That explicitly require the type and name to be kept.

I can open a question on R8 team about this, and come back here if you want their confirmation.

**Comment by tonihei** [comment]
Let me try the full mode too (that wasn't enabled in my test I believe)

**Comment by Tolriq** [comment]
@tonihei yes it works with:

```
# This method is called from native code
-keep, includedescriptorclasses class androidx.media3.decoder.ffmpeg.FfmpegAudioDecoder {
  private java.nio.ByteBuffer growOutputBuffer(androidx.media3.decoder.SimpleDecoderOutputBuffer, int);
}

```

**Comment by tonihei** [comment]
Great, will send a fix for this then! Thanks for your help :)

**Comment by Tolriq** [comment]
May I ask again some help on https://github.com/androidx/media/pull/369 then ? ;) Kills me to no be able to provide the test file to get this merged.

**Comment by tonihei** [comment]
> May I ask again some help on https://github.com/androidx/media/pull/369 then ?

I think this fell through the cracks a bit because we were waiting to see if there is a way to produce a test file for unit tests.

**Pull request merged**